src/box/sql/build.c

```diff
@@ -2030,49 +2031,104 @@ sql_create_foreign_key(struct Parse *parse_context)
 	goto exit_create_fk;
 }
 
-/**
- * Emit code to drop the entry from _index or _ck_contstraint or
- * _fk_constraint space corresponding with the constraint type.
- */
 void
-sql_drop_constraint(struct Parse *parse_context)
+sql_drop_table_constraint(struct Parse *parser, const struct Token *table_name,
+			  const struct Token *name)
 {
-	struct drop_entity_def *drop_def =
-		&parse_context->drop_constraint_def.base;
-	assert(drop_def->base.entity_type == ENTITY_TYPE_CONSTRAINT);
-	assert(drop_def->base.alter_action == ALTER_ACTION_DROP);
-	const char *table_name = drop_def->base.entity_name->a[0].zName;
-	assert(table_name != NULL);
-	struct space *space = space_by_name0(table_name);
+	const struct space *space = sql_space_by_token(table_name);
 	if (space == NULL) {
-		diag_set(ClientError, ER_NO_SUCH_SPACE, table_name);
-		parse_context->is_aborted = true;
+		const char *table_name_str = sql_tt_name_from_token(table_name);
+		diag_set(ClientError, ER_NO_SUCH_SPACE, table_name_str);
+		parser->is_aborted = true;
 		return;
 	}
-	char *name = sql_normalized_name_region_new(&parse_context->region,
-						    drop_def->name.z,
-						    drop_def->name.n);
-	struct Vdbe *v = sqlGetVdbe(parse_context);
-	assert(v != NULL);
-	struct constraint_id *id = space_find_constraint_id(space, name);
-	if (id == NULL) {
+	struct Vdbe *v = sqlGetVdbe(parser);
+	char *name_str = sql_name_from_token(name);
+
+	const struct tuple_constraint_def *fk =
+		sql_tuple_fk_by_token(space, name);
+	if (fk != NULL) {
 		sqlVdbeCountChanges(v);
-		sqlVdbeAddOp4(v, OP_DropTupleConstraint, space->def->id, 0, 0,
-			      sql_xstrdup(name), P4_DYNAMIC);
+		sqlVdbeAddOp4(v, OP_DropTupleForeignKey, space->def->id, 0, 0,
+			      name_str, P4_DYNAMIC);
 		return;
 	}
-	/*
-	 * We account changes to row count only if drop of
-	 * foreign keys take place in a separate
-	 * ALTER TABLE DROP CONSTRAINT statement, since whole
-	 * DROP TABLE always returns 1 (one) as a row count.
-	 */
-	assert(id->type == CONSTRAINT_TYPE_PK ||
-	       id->type == CONSTRAINT_TYPE_UNIQUE);
-	vdbe_emit_index_drop(parse_context, name, space->def,
-			     ER_NO_SUCH_CONSTRAINT, false);
-	sqlVdbeCountChanges(v);
-	sqlVdbeChangeP5(v, OPFLAG_NCHANGE);
+
+	const struct tuple_constraint_def *ck =
+		sql_tuple_ck_by_token(space, name);
+	if (ck != NULL) {
+		sqlVdbeCountChanges(v);
+		sqlVdbeAddOp4(v, OP_DropTupleCheck, space->def->id, 0, 0,
+			      name_str, P4_DYNAMIC);
+		return;
+	}
+
+	uint32_t index_id = sql_index_id_by_token(space, name);
+	if (index_id != UINT32_MAX) {
+		int regs = sqlGetTempRange(parser, 3);
+		sqlVdbeCountChanges(v);
+		sqlVdbeAddOp2(v, OP_Integer, space->def->id, regs);
+		sqlVdbeAddOp2(v, OP_Integer, index_id, regs + 1);
+		sqlVdbeAddOp3(v, OP_MakeRecord, regs, 2, regs + 2);
+		sqlVdbeAddOp3(v, OP_SDelete, BOX_INDEX_ID, regs + 2, 0);
+		sqlVdbeChangeP5(v, OPFLAG_NCHANGE);
+		sqlReleaseTempRange(parser, regs, 3);
+		return;
```

Comment:
Ditto.

Reply:
See above.

Reply:
Well, I see no `P4_DYNAMIC` here, so is there the similar machinery in scope of the emitter?

Reply:
This one is valid.

Reply:
Fixed.

Reply:
See no flaws now. Thanks!